internal/kustomization/build.go

```diff
@@ -0,0 +1,296 @@
+	kubeClient, err := utils.KubeClient(rcg)
+	if err != nil {
+		return nil, err
+	}
+
+	restMapper, err := rcg.ToRESTMapper()
+	if err != nil {
+		return nil, err
+	}
+
+	b := &Builder{
+		client:        kubeClient,
+		restMapper:    restMapper,
+		name:          name,
+		namespace:     *rcg.Namespace,
+		resourcesPath: resources,
+	}
+
+	for _, opt := range opts {
+		if err := opt(b); err != nil {
+			return nil, err
+		}
+	}
+
+	if b.timeout == 0 {
+		b.timeout = defaultTimeout
+	}
+
+	return b, nil
+}
+
+func (b *Builder) getKustomization(ctx context.Context) (*kustomizev1.Kustomization, error) {
+	namespacedName := types.NamespacedName{
+		Namespace: b.namespace,
+		Name:      b.name,
+	}
+
+	k := &kustomizev1.Kustomization{}
+	err := b.client.Get(ctx, namespacedName, k)
+	if err != nil {
+		return nil, err
+	}
+
+	return k, nil
+}
+
+// Build builds the yaml manifests from the kustomization object
+// and overlays the manifests with the resources specified in the resourcesPath
+// It expects a kustomization.yaml file in the resourcesPath, and it will
+// generate a kustomization.yaml file if it doesn't exist
+func (b *Builder) Build() ([]byte, error) {
+	m, err := b.build()
+	if err != nil {
+		return nil, err
+	}
+
+	resources, err := m.AsYaml()
+	if err != nil {
+		return nil, fmt.Errorf("kustomize build failed: %w", err)
+	}
+
+	return resources, nil
+}
+
+func (b *Builder) build() (m resmap.ResMap, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), b.timeout)
+	defer cancel()
+
+	// Get the kustomization object
+	k, err := b.getKustomization(ctx)
+	if err != nil {
+		return
+	}
+
+	// store the kustomization object
+	b.kustomization = k
+
+	// generate kustomization.yaml if needed
+	action, er := b.generate(*k, b.resourcesPath)
+	if er != nil {
+		errf := CleanDirectory(b.resourcesPath, action)
+		err = fmt.Errorf("failed to generate kustomization.yaml: %w", fmt.Errorf("%v %v", er, errf))
+		return
+	}
+
+	b.action = action
+
+	defer func() {
+		errf := b.Cancel()
+		if err == nil {
+			err = errf
+		}
+	}()
+
+	// build the kustomization
+	m, err = b.do(ctx, *k, b.resourcesPath)
+	if err != nil {
+		return
+	}
+
+	for _, res := range m.Resources() {
+		// set owner labels
+		err = b.setOwnerLabels(res)
+		if err != nil {
+			return
+		}
+
+		// make sure secrets are masked
+		err = trimSopsData(res)
+		if err != nil {
+			return
+		}
+	}
+
+	return
+
+}
+
+func (b *Builder) generate(kustomization kustomizev1.Kustomization, dirPath string) (action, error) {
+	data, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&kustomization)
+	if err != nil {
+		return "", err
+	}
+	gen := NewGenerator(unstructured.Unstructured{Object: data})
+
+	// acuire the lock
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	return gen.WriteFile(dirPath, WithSaveOriginalKustomization())
+}
+
+func (b *Builder) do(ctx context.Context, kustomization kustomizev1.Kustomization, dirPath string) (resmap.ResMap, error) {
```

Comment:
should this recursively build kustomization custom resources it finds in its path?

Reply:
similar concerns applies to the `Builder.Diff()` method

Reply:
The code rely on your `kustomization.yaml` if it does exist. Otherwise, it generates one by walking the directory specified by the `--path` flag and taking every `.yml`and `.yaml` into account. It all happens in `generateKustomization(dirPath string) (action, error)`, and it keeps(as its expected to) the kustomize-controller behavior.

Reply:
> The code rely on your kustomization.yaml if it does exist. Otherwise, it generates one by walking the directory specified by the --path flag and taking every .ymland .yaml into account

so I think we might be referring to different things here, `kustomization.yaml` is sort of overloaded, I am referring to the flux's kustomization custom resource and not kustomize's `kustomization.yaml`.

> It all happens in generateKustomization(dirPath string) (action, error), and it keeps(as its expected to) the kustomize-controller behavior.

I agree this the behavior of a single loop of kustomize-controller, however it is not the behavior for an entire changeset, for beyond one level deep tree of kustomization resource. In my use case I have in some cases 4 levels depth of kustomization resource i.e kustomization resource `a` creates and manages kustomization resource `ab`, `ac`, `ad`, kustomization resource `ab` creates and manages kustomization resource `abc`, `abd`, then `abc` and `abd` tracks regular kubernetes resources. If I introduce a new kustomization resource `abe`, this changeset will not detect the resources it manages if I point the flux cli(with this PR) at the file containing kustomization resource `a`, same issue if I delete kustomization resource `abd` while it has the `prune` flag set to true. The behavior I want will eventually happen across 3 runs of the kustomize-controller's reconcile loop... If the code here and in `Builder.Diff()` recurses through all the files then it can in one run show me what the kustomize-controller's reconcile loop would have done for the entirety of the changeset(inventory) it has been presented with.

Reply:
We rely on `kustomize` to build what's in the `path`. We recursively walk the path and included any sub-directory with a valid kustomization file. We don't discard any yaml file. As I said we do that only when no `kustomization.yaml` is detected. Otherwise we rely on the existing one to run `kustomize`. make sense?

Now we do not recursively apply the dry-run. But the reconcile dry-run is of your entire changeset because we have already recursively walked the path.

Reply:
>  But the reconcile dry-run is of your entire changeset because we have already recursively walked the path.

this is true if the directories are nested, the example I gave above and in my use case the directories are not nested

Reply:
oh ok. Do you want something like a dag, and to dry-run every step? Let's talk about this on slack or the related discussion.

Reply:
> Do you want something like a dag, and to dry-run every step?

there's an implicit dag, so what I was suggesting is that this build and diff walks the implicit dag and includes every subnode, so technically yes to your question :)

Reply:
> should this recursively build kustomization custom resources it finds in its path?

No as `flux diff kustomization` should do the same thing as the kustomize-controller.